bigtable/client/metadata_update_policy_test.cc

```diff
@@ -13,13 +13,278 @@
+  EXPECT_EQ(1, std::distance(result.first, result.second));
+
+  for (std::multimap<grpc::string_ref, grpc::string_ref>::iterator it =
+           result.first;
+       it != result.second; ++it) {
+    grpc::string_ref actual_x_goog_request_param = it->second;
+    EXPECT_EQ(expected_x_goog_request_param, actual_x_goog_request_param);
+  }
+  EXPECT_TRUE(wait_thread.joinable());
+  server->Shutdown();
+  wait_thread.join();
+}
+/// @test A test for setting metadata when table_id is known during the call.
+TEST(MetadataUpdatePolicy, RunWithEmbeddedServerLazyMetadata) {
+  std::unique_ptr<DefaultEmbeddedServer> server =
+      bigtable::internal::make_unique<DefaultEmbeddedServer>();
+  std::string address = server->address();
+  EXPECT_FALSE(server->address().empty());
+  bigtable::ClientOptions options;
+  options.set_admin_endpoint(address);
+  options.set_data_endpoint(address);
+  options.SetCredentials(grpc::InsecureChannelCredentials());
+  std::thread wait_thread([&server]() { server->Wait(); });
+  EXPECT_TRUE(wait_thread.joinable());
+
+  // Create the table against embedded server.
+  bigtable::TableAdmin admin(
+      bigtable::CreateDefaultAdminClient("fake-project", options),
+      "fake-instance");
+
+  admin.GetTable("fake-table-01");
+
+  grpc::string_ref expected_x_goog_request_param =
+      "name=projects/fake-project/instances/fake-instance/tables/fake-table-01";
+  // Get the metadata from embedded server
+  auto client_metadata = server->admin_service_client_metadata();
+  std::pair<std::multimap<grpc::string_ref, grpc::string_ref>::iterator,
+            std::multimap<grpc::string_ref, grpc::string_ref>::iterator>
+      result;
+  result = client_metadata.equal_range("x-goog-request-params");
+  // first check to see if client_metadata has only one occurance of parameter.
+  EXPECT_EQ(1, std::distance(result.first, result.second));
+
+  for (std::multimap<grpc::string_ref, grpc::string_ref>::iterator it =
+           result.first;
+       it != result.second; ++it) {
+    grpc::string_ref actual_x_goog_request_param = it->second;
+    EXPECT_EQ(expected_x_goog_request_param, actual_x_goog_request_param);
+  }
+  EXPECT_TRUE(wait_thread.joinable());
+  server->Shutdown();
+  wait_thread.join();
+}
+
+/// @test A test for setting metadata when table_id is known during the call.
+TEST(MetadataUpdatePolicy, RunWithEmbeddedServerParamTableName) {
+  std::unique_ptr<DefaultEmbeddedServer> server =
+      bigtable::internal::make_unique<DefaultEmbeddedServer>();
+  std::string address = server->address();
+  EXPECT_FALSE(server->address().empty());
+  bigtable::ClientOptions options;
+  options.set_admin_endpoint(address);
+  options.set_data_endpoint(address);
+  options.SetCredentials(grpc::InsecureChannelCredentials());
+  std::thread wait_thread([&server]() { server->Wait(); });
+  EXPECT_TRUE(wait_thread.joinable());
+
+  bigtable::Table table(bigtable::CreateDefaultDataClient(
```

Comment:
As an alternative, consider the `InProcessChannel`:

https://grpc.io/grpc/cpp/classgrpc_1_1_server.html

We would need to create a new `bigtable::DataClient` that users a pre-defined channel (maybe in the `testing/` directory?), but it would avoid all the mess with listening ports.

Reply:
created new DataClient and AdminClient to use predefined channel.